### ISSUE TITLE
System.Text.Json DateTime and DateTimeOffset only write six (6) significant digits for the seconds fraction

### ISSUE THREAD

**Issue by michaelstonis** [dotnet/runtime#91134]
### Description

When using a `DateTimeOffset` or `DateTime` and calling `ToString("O")` will result in a date time string with seven (7) significant digits in the seconds fraction.

The [documentation for System.TextJson](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.DateTimeOffset.cs#L21C17-L21C17) states that it uses the same "O" format and the example shown also has seven significant digits shown.

### Reproduction Steps

Simple console application.

```
using System.Text.Json;

var nowDto = DateTimeOffset.Now;
var nowDt = DateTime.Now;

var roundtripNowDtoFromToString = nowDto.ToString("O");
var roundtripNowDtoFromSystemTextJson = JsonSerializer.Serialize(nowDto).Replace("\"", string.Empty);

var roundtripNowDtFromToString = nowDt.ToString("O");
var roundtripNowDtFromSystemTextJson = JsonSerializer.Serialize(nowDt).Replace("\"", string.Empty);

Console.WriteLine($"DateTimeOffset - ToString: {roundtripNowDtoFromToString}");
Console.WriteLine($"DateTimeOffset - Json:     {roundtripNowDtoFromSystemTextJson}");

Console.WriteLine($"DateTime       - ToString: {roundtripNowDtFromToString}");
Console.WriteLine($"DateTime       - Json:     {roundtripNowDtFromSystemTextJson}");
```

This results in the following output:

```
DateTimeOffset - ToString: 2023-08-25T15:03:07.4735810-05:00
DateTimeOffset - Json:     2023-08-25T15:03:07.473581-05:00
DateTime       - ToString: 2023-08-25T15:03:07.5330680-05:00
DateTime       - Json:     2023-08-25T15:03:07.533068-05:00
```

### Expected behavior

When serializing a DateTimeOffset or DateTime the seconds fraction will have 7 decimal places.

### Actual behavior

When serializing a DateTimeOffset or DateTime the seconds fraction has 6 decimal places.

### Known Workarounds

If you specify the format specifically using `yyyy'-'MM'-'dd'T'HH':'mm':'ss.FFFFFFFK` it will be consistent.

### Configuration

- .NET 7
- macOS 13.5

### Other information

It appears that NewtonsoftJson has a similar formatting issue.

https://github.com/JamesNK/Newtonsoft.Json/issues/2780

**Comment by Clockwork-Muse** [comment]
.... wait till you find out about the deserialize/parse behavior....

**Comment by danmoseley** [comment]
Is there an accuracy issue here, or just aesthetic ? In your example, JSON case seems to be the shortest round trippable value, while the others are padding.

https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xAdACoCmAHggwFIDOA9jADcFCvgDMdQjQDCNAN4UaSyVIXllGmos1byFAG4RkNGHwDuAEQR8aAXhoWICFkyhgWAeQBmXni04AcubC5IbGppYIdg5OLm4sDEFmIQZGNMh86BgIyFAADklWfABiGWBMfADKObAA5tERRcxVNTC1ABQARB6dAJQhYemZMNm5BeZFpXxg9KwcvALRCzCVLLkQADZQAF5rDKvrW7vtjda9DABKLHkbEADGLF0AOp2dcHTEVAwAomB5CABPfoiUJpDJZGrjSJTcotXJtBoTTgVarwjrdPoDMHDUb5QoIGGzdgIZZLfgrNZQTY7PYHKlHR6nc5XG73R6dF5vD5fX7/IEpSjEACc7QAJJ1HM5XO5vL5/DQALQ0FGtWogeTgkaQ/ElMoqtEAX0xomFYolsWlnh8fiiSuW6o0ck1uKhkzKRPm5KNwL0gpF4slcXcOkVyrhdXVTpx2qRMP1dW9IXo/vNUviIbt5IdyijELG+MJxDmJK9mN0IfLIe0SgNFANQA===

**Comment by michaelstonis** [comment]
The issue is really just around consistency.

I encountered this issue when building a system that persists data to a database and compares data at rest. I noticed that when `DateTime/Offset` objects were directly stored, the value was slightly different than when they were contained in JSON serialized by System.Text.Json. This is easily fixable by either updating the format of `ToString` or adding a custom converter for the JSON serializer. Understandably, the format could be different for a serialized JSON object, but the documentation is misleading.

Below are a few thoughts on it:

1. The .NET `DateTime/Offset` definition of the "O" formatter or the round-trip format or .NET version of ISO 8601 implementation is standardized across the platform. That means if something from .NET is serialized to that format, it can always be compared consistently without additional parsing or manipulation.
2. The docs for the System.Text.Json documentation are updated to say that it uses the `yyyy'-'MM'-'dd'T'HH':'mm':'ss.FFFFFFFK` format. This removes any ambiguity about it being directly comparable to the "O" format used by `DateTime/Offset`.

**Comment by danmoseley** [comment]
> if something from .NET is serialized to that format, it can always be compared consistently without additional parsing or manipulation.

In general comparing numbers as strings is not ideal. I don't see a good reason why they should use the same string representation, so long as they're accurate and roundtrippable. However it's up to JSON area owners here.

**Comment by michaelstonis** [comment]
I agree entirely that comparing numbers as strings is in no way ideal. 

The case I am trying to make is that if something states that it _does_ serialize to a known format and the documentation supports that, it should do what it says on the box or the docs should be updated.

In this case, the strings `2023-08-25T15:03:07.4735810-05:00` and `2023-08-25T15:03:07.473581-05:00` are not equal. It is an easily solvable problem in more than a few ways, but a direct comparison cannot be done without parsing.

**Comment by yufeih** [comment]
Keeping the zeros makes the string lexicographical comparable which is an important feature of ISO date strings. With this, it is a lot easier for JavaScript to compare dates in JSON date responses without having to parse the string as `Date`.

**Comment by markusschaber** [comment]
On the other hand, I like the fact that it shortens trailing zeroes. If your data only cares about seconds, and doesn't have sub second fractions, you don't want to have ".0000000" appended to every timestamp.
